Todo app week 2

sapphire2359 opened the pull request:
Adding changes made to the Todo App for week 2

Added code for the following sections
[1] Advanced functionality
[2] API
[3] Web Page

prakashg2359 committed:
main file and json file

prakashg2359 committed:
json file moved to "Using JSON/data" folder

prakashg2359 committed:
Changes according to feedback - [1] "Id" field added in todo item, status - "not started", "started", "complete" declared as constants,  main put at top of file followed by other functions

prakashg2359 committed:
This is the To Do App branch for week 2

sapphire2359 committed:
Create README.md

Adding how to use information for Todo app CLI, test API and static and dynamic web page

sapphire2359 committed:
Update README.md

AdamGreenhillBJSS commented on README.md:
```diff
@@ -0,0 +1,82 @@
+
+Open a browser and input the following url.
+
+http://localhost:8080/static/
```
should be this to actually see the HTML in the browser
```suggestion
http://localhost:8080/static/about.html
```

AdamGreenhillBJSS commented on README.md:
```diff
@@ -0,0 +1,82 @@
+
+[4] Delete - /delete
+
+curl http://localhost:8080/delete?id=1
```
I get an error when running this - does it need to look more like the other commands with a verb and a JSON payload?
> no matches found: http://localhost:8080/delete?id=12

AdamGreenhillBJSS commented on todoapp-json/api/api.go:
```diff
@@ -0,0 +1,137 @@
+
+	// Serve dynamic list page
+	mux.Handle("/list", TraceMiddleware(http.HandlerFunc(listHandler)))
+}
```
Just a style note, but this kind of setup code would usually be in your `main` function/file. It makes it easier to quickly see what routes have been defined, as soon as you open up the project code. The actual handler functions can remain in this `api.go` file.

AdamGreenhillBJSS commented:
Can you take a look at your use of logging? This is the output from my test run of the program and there are a couple things that don't look right.

1. Trace ID is always the same
2. There is this `...=...` part cluttering up the List log entries

```
time=2025-06-16T09:23:33.930+01:00 level=INFO msg="Server started on :8080"
time=2025-06-16T09:27:47.465+01:00 level=INFO msg="List all todo items..." traceID=trace-9999 ...=...
time=2025-06-16T09:33:24.169+01:00 level=INFO msg="Added new todo item..." traceID=trace-9999 id=13
time=2025-06-16T09:33:24.171+01:00 level=INFO msg="Created todo" traceID=trace-9999 desc="Go for a run"
time=2025-06-16T09:34:34.926+01:00 level=INFO msg="Updated todo item" traceID=trace-9999 id=3
time=2025-06-16T09:37:54.852+01:00 level=INFO msg="List all todo items..." traceID=trace-9999 ...=...
```

AdamGreenhillBJSS commented on todoapp-json/logic/logic.go:
```diff
@@ -0,0 +1,84 @@
+	if !models.IsValidStatus(status) {
+		return errors.New("invalid status")
+	}
+	//id := len(todos) + 1
```
can you remove commented out code if it's not needed please? There are a few other places too, please check with a text search or something

AdamGreenhillBJSS commented on todoapp-json/logic/logic.go:
```diff
@@ -0,0 +1,84 @@
+				//status validation
+				if !models.IsValidStatus(newStatus) {
+					return errors.New("invalid status")
+				}
```
I would suggest putting this validation right at the point where you parse in the request, and then you can return a HTTP 400 Bad Request if the check fails

AdamGreenhillBJSS commented on todoapp-json/logic/logic.go:
```diff
@@ -0,0 +1,84 @@
+			continue
+		}
+		newTodoItems = append(newTodoItems, todoItem)
+	}
```
You could use `slices.Index` and `slices.Delete` here. It probably does the same thing under the hood to be fair

sapphire2359 committed:
Update README.md

changed the delete command that was written - curl http://localhost:8080/delete?id=16.

the "-X" and "DELETE" have to be added  otherwise the default HTTP method GET will be called. 

For that reason the previous delete command will give error.

sapphire2359 committed:
Update README.md

For /update -> the HTTP method should be "PUT" instead of "POST"

prakashg2359 committed:
Feedback changes and other changes for week 2

prakashg2359 committed:
final changes for week 2

sapphire2359 committed:
Update README.md

How to run test instruction added for - Logic test

sapphire2359 committed:
Update README.md

correcting format changes from previous commit

sapphire2359 committed:
Update README.md

minor correction
